pkg/endpoint/bpf.go

```diff
@@ -679,11 +679,68 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 	defer e.desiredPolicy.Ready()
 
+	var (
+		desiredRedirects map[string]uint16
+		finalizeFunc     revert.FinalizeFunc
+		revertFunc       revert.RevertFunc
+	)
+	// Get currently realized redirects
+	previouslyRealizedRedirects := e.GetRealizedRedirects()
+
 	// Apply pending policy map changes so that desired map is up-to-date before
-	// we update network policies and sync the maps below.
-	err = e.applyPolicyMapChanges()
-	if err != nil {
-		return err
+	// syncing the maps below.
+	if e.SecurityIdentity != nil {
+		_ = e.updateAndOverrideEndpointOptions(nil)
+
+		// Walk the L4Policy to add new redirects and update the desired policy for existing
+		// redirects.  Do this before updating the bpf policy maps, so that the proxies are
+		// ready when new traffic is redirected to them.
+		if e.desiredPolicy != nil && !e.isProperty(PropertySkipBPFPolicy) {
+			stats.proxyConfiguration.Start()
+			// Deny policies do not support redirects
+			desiredRedirects, err, finalizeFunc, revertFunc = e.addNewRedirects(datapathRegenCtxt.proxyWaitGroup)
+			stats.proxyConfiguration.End(err == nil)
+			if err != nil {
+				return err
+			}
+			datapathRegenCtxt.finalizeList.Append(finalizeFunc)
+			datapathRegenCtxt.revertStack.Push(revertFunc)
+
+			// At this point desiredRedirects contains the new realized redirects,
+			// so update e.realizedRedirects pointer.
+			// Note that this needs to be called before calling syncPolicyMap below,
+			// as the set of realized redirects is needed there.
+			e.getLogger().Debugf("Storing desiredRedirects: %v", desiredRedirects)
+			e.realizedRedirects.Store(&desiredRedirects)
+			datapathRegenCtxt.revertStack.Push(func() error {
+				// revert back to the set of previously realized redirects
+				e.realizedRedirects.Store(&previouslyRealizedRedirects)
+				return nil
+			})
+		}
+
+		// Apply incremental changes to desiredPolicy and Configure the new network policy
+		// with the proxies.
+		//
+		// If we have a new policy, it is likely that incremental updates were applied to
+		// the old policy while we were waiting for the endpoint lock. If we were to sync
+		// policy maps without applying incremental updates first we could be flapping
+		// network policy backwards just to be updating it again after applying incremental
+		// updates later.
+		//
+		// This must be done after adding new redirects, as waiting for policy update
+		// ACKs is disabled when there are no listeners, which is the case before the first
+		// redirect is added.
+		//
+		// Do this before updating the bpf policy maps (later), so that the proxy listeners
+		// have a chance to be ready when new traffic is redirected to them.  Note that it
+		// is possible for further incremental changes to be applied before and after the
+		// bpf policy maps have been synchronized for the new policy.
+		//
+		err = e.applyPolicyMapChanges(regenContext, e.desiredPolicy != e.realizedPolicy)
+		if err != nil {
+			return err
+		}
```

Comment:
1st commit only refactors operations happening within the same endpoint lock, so there is no locking impact. It moves calls to `e.updateAndOverrideEndpointOptions(nil)` up so that it can be called from one code path rather than many. `addNewRedirects` is moved to happen before `applyPolicyMapChanges` so that the proxy ports exist when creating redirect map state entries. Some logging is moved to `applyPolicyMapChanges`, still under the same conditions. `updateNetworkPolicy` is moved into `applyPolicyMapChanges`, to the end, so that it still happens after the desiredPolicy has been updated by `ConsumeMapChanges` before the proxy network policy is updated.

There is one possibly observable change in the 1st commit: The bpf map updates via `applyPolicyMapChanges` are skipped, and are instead done by the following full map sync when a new policy has been computed. Previously is was possible that e.g., new redirects were pushed to the bpf maps by `applyPolicyMapChanges` when Envoy was still processing the new listeners and/or the network policy for them. This could have resulted in "policy not found" type of errors in the proxy, if traffic was redirected to the (new) listeners for endpoints whose policy was not yet updated.

The main change is in the 2nd commit, where `addNewRedirects` is computed earlier, between the selector policy update and creation of the new map state. Just checked for locking requirements, and I was missing the endpoint lock there! It looks like read lock suffices, though. This code path calls in to `pkg/proxy`, which may call back to `endpoint` via `EndpointUpdater`.

Reply:
That sounds reasonable to me; this would be great to include in the commit msg itself :)